tutorials/application_design/c++11/subscriber.cxx

```diff
@@ -0,0 +1,254 @@
+
+    for (;;) {
+        waitset.dispatch();
+        std::lock_guard<std::mutex> lock(mutex);
+        display_condition.extensions().trigger_value(false);
+    }
+}
+
+void SubscriberDashboard::display_app()
+{
+    std::stringstream ss;
+    auto now = std::chrono::system_clock::now();
+    ss << "[[ DASHBOARD: " << now.time_since_epoch().count() << " ]]\n";
+    {
+        auto online = online_vehicles();
+        ss << "Online vehicles: " << online.size() << "\n";
+        for (auto &item : online) {
+            ss << "- Vehicle " << item.vin << ":\n";
+            ss << "  Fuel updates: " << item.fuel_history.size() << "\n";
+            ss << "  Last known destination: "
+               << (item.current_destination
+                           ? utils::to_string(*item.current_destination)
+                           : "None")
+               << "\n";
+            ss << "  Last known fuel level: "
+               << (item.fuel_history.empty()
+                           ? "None"
+                           : utils::to_string(item.fuel_history.back()))
+               << "\n";
+        }
+    }
+    {
+        auto offline = offline_vehicles();
+        ss << "Offline vehicles: " << offline.size() << "\n";
+        for (auto &item : offline) {
+            ss << "- Vehicle " << item.vin << ":\n";
+            ss << "  Mean fuel consumption: "
+               << std::accumulate(
+                          item.fuel_history.begin(),
+                          item.fuel_history.end(),
+                          0.0)
+                            / item.fuel_history.size()
+               << "\n";
+            ss << "  Known reached destinations: "
+               << item.reached_destinations.size() << "\n";
+            for (auto &destination : item.reached_destinations) {
+                ss << "    - " << utils::to_string(destination) << "\n";
+            }
+        }
+    }
+
+    std::cout << ss.str() << std::endl;
+}
+
+void SubscriberDashboard::metrics_app()
+{
+    for (const auto &sample : metrics_reader_.take()) {
+        auto it = dashboard_data_.find(sample.info().instance_handle());
+        // If not a tracked vehicle, track it.
+        if (it == dashboard_data_.end()) {
+            if (!sample.info().valid())
+                continue;
+
+            auto new_handle = sample.info().instance_handle();
+            auto new_data = DashboardItem { sample.data().vehicle_vin() };
+            it = dashboard_data_.emplace(new_handle, new_data).first;
+        }
+
+        auto &item = it->second;
+        item.is_historical = sample.info().state().instance_state()
+                != dds::sub::status::InstanceState::alive();
+
+        if (!sample.info().valid() && item.is_historical) {
+            continue;
+        }
+
+        item.fuel_history.push_back(sample.data().fuel_level());
+    }
+}
+
+void SubscriberDashboard::transit_app()
+{
+    for (const auto &sample : transit_reader_.take()) {
```

Comment:
Have you considered using the DataReader cache (in a similar way as the Last-Value Cache module) to keep track of online vehicles and model offline vehicles as a dispose?

I think that design aligns much better with Connext best practices/design patterns. 

If you already thought about it, was there any major blocker that made you keep the information at the application level?

Reply:
I did not think about using the Reader queue to hold the internal state, as seen also in the Python example.

I'm not sure if redo-ing the example using `read` will be straightforward, so perhaps we could use this as an excuse to redirect readers to Last-Value Cache.

Reply:
I would like to evaluate this design a little more. I'm OK with doing that outside this PR.